Client.py

```diff
@@ -0,0 +1,227 @@
+                    "%s 200 OK\r\n"
+                    "Content-Length: %s\r\n"
+                    "Content-Type: %s\r\n"
+                    "\r\n"
+                ) % (
+                    constants.HTTP_VERSION,
+                    self.file.get_file_size(),
+                    MIME_MAPPING.get(
+                        os.path.splitext(
+                            self.request.get_uri()
+                        )[1].lstrip('.'),
+                        'application/octet-stream',
+                    ),
+                )
+                self.logger.debug("Created file and request")
+                if len(parsed_lines) == 1:
+                    self._recv_buff = ""
+                else:
+                    self._recv_buff = parsed_lines[1]
+
+            # If we do have request line, get headers
+            if self.request is not None:
+                if not self.request.full_request and \
+                        constants.DOUBLE_CRLF in self._recv_buff:
+                    self.request.full_request = True
+                self._recv_lines()
+                self.logger.debug("after recv lines %s" % self._recv_buff)
+                if constants.CRLF in self._recv_buff:
+                    lines = self._recv_buff.split(constants.DOUBLE_CRLF)[
+                        0].split(constants.CRLF)
+                    for line in lines[:-1]:
+                        if ": " in line:
+                            self._parse_header(line)
+                    if ": " in lines[-1]:
+                        self._parse_header(lines[-1])
+                        self._recv_buff = ""
+                    else:
+                        self._recv_buff = lines[-1]
+
+            self.logger.debug("Now recv_buff is %s" % self._recv_buff)
+
+        except OSError as e:
+            self.logger.error('Exception ', exc_info=True)
+            if e.errno == errno.ENOENT:
+                self._send_buff = util.creat_error(404, 'File Not Found', e)
+                self.request = Request.Request()
+                self.request.full_request = True
+                self._recv_buff = ""
+            else:
+                self._send_buff = util.creat_error(500, 'Internal Error', e)
+                self.request = Request.Request()
+                self.request.full_request = True
+                self._recv_buff = ""
+
+    def send(self):
+        """ Fill self.send_buff with all the data it needs and then send it
+        """
+        free_space_in_buffer = self._buff_size - len(self._send_buff)
+        if free_space_in_buffer > 0 and self.file is not None:
+            self._send_buff += self.file.read_buff(free_space_in_buffer)
+        if len(self._send_buff) > 0:
+            maybe_sent_status = False
+            if constants.DOUBLE_CRLF in self._send_buff:
+                maybe_sent_status = True
+            self._send_my_buff()
+            if constants.DOUBLE_CRLF in self._send_buff:
+                maybe_sent_status = False
```

Comment:
I do not understand this! do you try to guess if you sent status line based on parsing your own response?!?!